src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,365 @@
+                    None => {
+                        return Err(Error::new(
+                            ErrorKind::Other,
+                            "no attestation key ID provided for setting Init Quote Ex request",
+                        ));
+                    }
+                };
+                let size = match size {
+                    Some(s) => s,
+                    None => {
+                        return Err(Error::new(
+                            ErrorKind::Other,
+                            "no key size provided for setting Init Quote Ex request",
+                        ));
+                    }
+                };
+                let mut msg = Request_InitQuoteExRequest::new();
+                msg.set_timeout(TIMEOUT);
+                msg.set_b_pub_key_id(true);
+                msg.set_att_key_id(akid);
+                msg.set_buf_size(size as u64);
+                req.set_initQuoteExReq(msg);
+                Ok(req)
+            }
+            ReqType::KeySize => {
+                let akid = match akid {
+                    Some(a) => a,
+                    None => {
+                        return Err(Error::new(
+                            ErrorKind::Other,
+                            "no attestation key ID provided for setting Init Quote Ex request to get key size",
+                        ));
+                    }
+                };
+                let mut msg = Request_InitQuoteExRequest::new();
+                msg.set_timeout(TIMEOUT);
+                msg.set_b_pub_key_id(false);
+                msg.set_att_key_id(akid);
+                req.set_initQuoteExReq(msg);
+                Ok(req)
+            }
+            ReqType::Quote => {
+                let report = match report {
+                    Some(r) => r,
+                    None => {
+                        return Err(Error::new(
+                            ErrorKind::Other,
+                            "no Report provided for setting Get Quote Ex request",
+                        ));
+                    }
+                };
+                let akid = match akid {
+                    Some(a) => a,
+                    None => {
+                        return Err(Error::new(
+                            ErrorKind::Other,
+                            "no attestation key ID provided for setting Get Quote Ex request",
+                        ));
+                    }
+                };
+                let mut msg = Request_GetQuoteExRequest::new();
+                msg.set_timeout(TIMEOUT);
+                msg.set_report(report.to_vec());
+                msg.set_att_key_id(akid);
+                msg.set_buf_size(SGX_QUOTE_SIZE as u32);
+                req.set_getQuoteExReq(msg);
+                Ok(req)
+            }
+        }
+    }
+
+    fn send_request(&self, req: Request, mut stream: UnixStream) -> Result<Response, Error> {
+        // Set up writer
+        let mut buf_wrtr = vec![0u8; size_of::<u32>()];
```

Comment:
nit; this could be just `buf`

Reply:
This is copied code from https://github.com/enarx/enarx/blob/7f691c91783c841c807d7fc72c8038a2210d8007/src/backend/sgx/attestation.rs
don't blame this code to @jarkkojs

Reply:
`Write` trait is not implemented for array, e.g.

```
-        let mut buf_wrtr = vec![0u8; size_of::<u32>()];
+        let mut buf_wrtr: &[u8] = &[0, 0, 0, 0];
```

will result:

```
error[E0277]: the trait bound `&[u8]: std::io::Write` is not satisfied
   --> src/backend/sgx/attestation.rs:132:35
    |
132 |         match req.write_to_writer(&mut buf_wrtr) {
    |                                   ^^^^^^^^^^^^^ the trait `std::io::Write` is not implemented for `&[u8]`
    |
    = help: the following implementations were found:
              <&mut [u8] as std::io::Write>
    = note: `std::io::Write` is implemented for `&mut [u8]`, but not for `&[u8]`
    = note: required for the cast to the object type `dyn std::io::Write`
```

Reply:
For the record, same result comes with `let mut buf_wrtr: [u8; 4] = [0, 0, 0, 0];`.